lightningd/log.c

```diff
@@ -535,26 +615,41 @@ char *arg_log_to_file(const char *arg, struct lightningd *ld)
 	if (size > 0)
 		fprintf(ld->log->lr->outf, "\n\n\n\n");
 
-	/* Catch up */
-	list_for_each(&ld->log->lr->log, i, list)
-		maybe_print(ld->log, i, 0);
-
 	log_debug(ld->log, "Opened log file %s", arg);
 	return NULL;
 }
 
 void opt_register_logging(struct lightningd *ld)
 {
 	opt_register_early_arg("--log-level",
-			       arg_log_level, show_log_level, ld->log,
-			       "log level (io, debug, info, unusual, broken)");
+			       opt_log_level, show_log_level, ld->log,
+			       "log level (io, debug, info, unusual, broken) [:prefix]");
 	opt_register_early_arg("--log-prefix", arg_log_prefix, show_log_prefix,
 			       ld->log,
 			       "log prefix");
 	opt_register_early_arg("--log-file=<file>", arg_log_to_file, NULL, ld,
 			       "log to file instead of stdout");
 }
 
+void logging_options_parsed(struct log_book *lr)
+{
+	const struct log_entry *l;
+
+	/* If they didn't set an explicit level, set to info */
+	if (!lr->default_print_level) {
+		lr->default_print_level = tal(lr, enum log_level);
+		*lr->default_print_level = DEFAULT_LOGLEVEL;
+	}
+
+	/* Catch up, since before we were only printing BROKEN msgs */
+	list_for_each(&lr->log, l, list) {
+		if (l->level >= filter_level(lr, l->prefix))
+			log_to_file(l->prefix, l->level, l->node_id, false,
+				    &l->time, l->log,
+				    l->io, tal_bytelen(l->io), lr->outf);
+	}
+}
```

Comment:
It always could; generally BROKEN messages are fatal though, as you anticipated.

I figured the duplication here is clearer, since now they'll be in context.